File: DailyChallenges/Scala/LeetCode_652.scala
object Solution {
    def findDuplicateSubtrees(root: TreeNode): List[TreeNode] = {
        val m = collection.mutable.HashMap.empty[String, Int]
        var l = List.empty[TreeNode]
        def dfs(node: TreeNode): String = if (node == null) return "" else {
            val s = "(" + dfs(node.left) + "," + node.value + "," + dfs(node.right) + ")";
            m(s) = m.getOrElse(s, 0) + 1
            if (m(s) == 2) l = node :: l
            s
        } 
        dfs(root)
        l
    }
}

File: DailyChallenges/CPP/LeetCode_652.cpp
class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        unordered_map<string, int> m;
        vector<TreeNode*> ans;
        function<string(TreeNode*)> dfs = [&](TreeNode* node){
            if (!node) return string("");
            string s = "(" + dfs(node->left) + "," + to_string(node->val) + "," + dfs(node->right)+ ")";
            if (m[s]++ == 1) ans.push_back(node);
            return s;
        };
        dfs(root);
        return ans;
    }
};
